File: broke/src/api/gameSocial.ts
import { ACCESS_TOKEN_KEY } from './config'
import { requestJson } from './http'

function authHeaders(): HeadersInit {
  const token = typeof window !== 'undefined' ? localStorage.getItem(ACCESS_TOKEN_KEY) : null
  if (!token) throw new Error('로그인이 필요합니다.')
  return { Authorization: `Bearer ${token}` }
}

export type OnlinePeer = {
  id: number
  nickname: string
  has_game_location: boolean
}

export type GameOnlineSummary = {
  my_location_set: boolean
  my_game_lat?: number | null
  my_game_lng?: number | null
  online_peers: OnlinePeer[]
}

export type PeerLocation = {
  id: number
  nickname: string
  lat: number
  lng: number
}

export type CloserRound = {
  peer_a: { id: number; nickname: string }
  peer_b: { id: number; nickname: string }
}

export type CloserGuessResult = {
  correct: boolean
  closer_user_id: number
  distances_km: Record<string, number>
  markers: Array<{
    id: number
    nickname: string
    lat: number
    lng: number
    kind: string
  }>
}

export async function postGamePresence(): Promise<void> {
  await requestJson<void>('/game/presence', {
    method: 'POST',
    headers: authHeaders(),
  })
}

export async function postGameLocation(lat: number, lng: number): Promise<void> {
  await requestJson<void>('/game/location', {
    method: 'POST',
    headers: authHeaders(),
    body: JSON.stringify({ lat, lng }),
  })
}

export async function fetchGameOnlineSummary(): Promise<GameOnlineSummary> {
  return requestJson<GameOnlineSummary>('/game/online-summary', {
    headers: authHeaders(),
  })
}

export async function fetchPeerLocation(peerId: number): Promise<PeerLocation> {
  return requestJson<PeerLocation>(`/game/peer-location/${peerId}`, {
    headers: authHeaders(),
  })
}

export async function fetchCloserRound(): Promise<CloserRound> {
  return requestJson<CloserRound>('/game/closer-round', {
    headers: authHeaders(),
  })
}

export async function postCloserGuess(
  peerAId: number,
  peerBId: number,
  pickedUserId: number,
): Promise<CloserGuessResult> {
  return requestJson<CloserGuessResult>('/game/closer-guess', {
    method: 'POST',
    headers: authHeaders(),
    body: JSON.stringify({
      peer_a_id: peerAId,
      peer_b_id: peerBId,
      picked_user_id: pickedUserId,
    }),
  })
}

File: gourmet/app/schemas/game_social.py
from __future__ import annotations

from pydantic import BaseModel, Field


class GameLocationIn(BaseModel):
    lat: float = Field(ge=-90, le=90)
    lng: float = Field(ge=-180, le=180)


class OnlinePeerOut(BaseModel):
    id: int
    nickname: str
    has_game_location: bool


class GameOnlineSummaryOut(BaseModel):
    my_location_set: bool
    my_game_lat: float | None = None
    my_game_lng: float | None = None
    online_peers: list[OnlinePeerOut]


class PeerLocationOut(BaseModel):
    id: int
    nickname: str
    lat: float
    lng: float


class CloserPeerHintOut(BaseModel):
    id: int
    nickname: str


class CloserRoundOut(BaseModel):
    peer_a: CloserPeerHintOut
    peer_b: CloserPeerHintOut


class CloserGuessIn(BaseModel):
    peer_a_id: int = Field(ge=1)
    peer_b_id: int = Field(ge=1)
    picked_user_id: int = Field(ge=1)


class CloserGuessOut(BaseModel):
    correct: bool
    closer_user_id: int
    distances_km: dict[str, float]
    markers: list[dict[str, float | int | str]]
